agent/testflinger_agent/tarfile_patch.py

```diff
@@ -0,0 +1,265 @@
+    # Ensure we stay in the destination
+    target_path = os.path.realpath(os.path.join(dest_path, name))
+    if os.path.commonpath([target_path, dest_path]) != dest_path:
+        raise OutsideDestinationError(member, target_path)
+    # Limit permissions (no high bits, and go-w)
+    mode = member.mode
+    if mode is not None:
+        # Strip high bits & group/other write bits
+        mode = mode & 0o755
+        if for_data:
+            # For data, handle permissions & file types
+            if member.isreg() or member.islnk():
+                if not mode & 0o100:
+                    # Clear executable bits if not executable by user
+                    mode &= ~0o111
+                # Ensure owner can read & write
+                mode |= 0o600
+            elif member.isdir() or member.issym():
+                # Ignore mode for directories & symlinks
+                mode = None
+            else:
+                # Reject special files
+                raise SpecialFileError(member)
+        if mode != member.mode:
+            new_attrs["mode"] = mode
+    if for_data:
+        # Ignore ownership for 'data'
+        if member.uid is not None:
+            new_attrs["uid"] = None
+        if member.gid is not None:
+            new_attrs["gid"] = None
+        if member.uname is not None:
+            new_attrs["uname"] = None
+        if member.gname is not None:
+            new_attrs["gname"] = None
+        # Check link destination for 'data'
+        if member.islnk() or member.issym():
+            if os.path.isabs(member.linkname):
+                raise AbsoluteLinkError(member)
+            if member.issym():
+                target_path = os.path.join(
+                    dest_path, os.path.dirname(name), member.linkname
+                )
+            else:
+                target_path = os.path.join(dest_path, member.linkname)
+            target_path = os.path.realpath(target_path)
+            if os.path.commonpath([target_path, dest_path]) != dest_path:
+                raise LinkOutsideDestinationError(member, target_path)
+    return new_attrs
+
+
+_KEEP = object()
+
+
+def replace(
+    tarinfo,
+    *,
+    name=_KEEP,
+    mtime=_KEEP,
+    mode=_KEEP,
+    linkname=_KEEP,
+    uid=_KEEP,
+    gid=_KEEP,
+    uname=_KEEP,
+    gname=_KEEP,
+    deep=True,
+    _KEEP=_KEEP,
+):
+    """Return a deep copy of self with the given attributes replaced."""
+    if deep:
+        result = copy.deepcopy(tarinfo)
+    else:
+        result = copy.copy(tarinfo)
+    if name is not _KEEP:
+        result.name = name
+    if mtime is not _KEEP:
+        result.mtime = mtime
+    if mode is not _KEEP:
+        result.mode = mode
+    if linkname is not _KEEP:
+        result.linkname = linkname
+    if uid is not _KEEP:
+        result.uid = uid
+    if gid is not _KEEP:
+        result.gid = gid
+    if uname is not _KEEP:
+        result.uname = uname
+    if gname is not _KEEP:
+        result.gname = gname
```

Comment:
```suggestion
    attributes = {
        "name": name,
        "mtime": mtime,
        "mode": mode,
        "linkname": linkname,
        "uid": uid,
        "gid": gid,
        "uname": uname,
        "gname": gname,
    }

    for key, value in attributes.items():
        if value is not _KEEP:
            setattr(result, key, value)
```

Reply:
Same here, this is copied from the standard library implementation of `tarfile`.